---
title: "Proyecto_driver_detection_dndscv"
output:
  html_document:
    df_print: paged
editor_options:
  markdown:
    wrap: sentence
---

```{r}
rm(list= ls())
setwd("/Users/celiamp/Documents/SEMESTRE 4/cancer/cancer_genomics")
library(dndscv)
muts = read.table("data_mutations.txt", header=T, sep="\t", stringsAsFactors=F)
head(muts)
```

```{r}
# Transformamos tabla que esta en formato MAF a lo que espera dndscv()
muts_dndscv <- data.frame(
  sampleID = muts$Tumor_Sample_Barcode,
  chr      = muts$Chromosome,
  pos      = muts$Start_Position,
  ref      = muts$Reference_Allele,
  mut      = muts$Tumor_Seq_Allele2
)

# Eliminar indels (solo dejar SNVs)
#muts_dndscv <- muts_dndscv[
#  nchar(muts_dndscv$ref) == 1 & nchar(muts_dndscv$mut) == 1, 
#]

```

```{r}
# To count the number of samples 
length(unique(muts_dndscv$sampleID))
```

```{r}
# And to get the number of mutations
nrow(muts_dndscv)
```


To see the mutational burden per sample we can do the following:

```{r}
barplot(sort(table(muts_dndscv$sampleID)), ylab="Number of mutations",
xlab="Donors", las=2, names.arg="")
```

## Are there any hypermutators in the cohort?

\*t is relevant to explore our data in this way because hypermutators can have a negative impact on the statistical power to detect drivers and because some hypermutators (e.g. POLE mutant tumors) are under mutational processes not properly modeled by a trinucleotide-substitution model.
Although there is no exact definition of a hypermutator, ***usually having more than 500 mutations in the exome (approximately 5.6810\^{4} mutations in the whole genome) can be considered a hypermutator.***

In general, it is good to exclude these samples from this analysis.

```{r}
# Identificar hipermutadores (más de 500 mutaciones)
# Contar mutaciones por muestra
mut_count <- table(muts_dndscv$sampleID)

# Identificar muestras con más de 500 mutaciones (hipermutadores)
hipermutadores <- names(mut_count[mut_count > 500])

# Mostrar resultados
cat("Muestras hipermutadoras:", hipermutadores, "\n")
```

## Driver detection

Here is the translated and slightly polished version in English:

---

**Now we will run the dNdScv analysis to detect genes under positive selection, filtering out hypermutated samples and saving the output matrices.**

**IMPORTANT NOTE!!!**
In the following code, the parameter `max_coding_muts_per_sample` was changed from 50 to 500 because, when set to 50, only one significant gene (TP53) was detected — which may not be accurate.

```{r}
# analyses of selection using the dNdScv and dNdSloc models
dout = dndscv(muts_dndscv,
              max_muts_per_gene_per_sample=3,
              max_coding_muts_per_sample= 500, #max_coding_muts_per_sample=50,
              outmats=T)
```


```{r}
# Looking at the output
names(dout)
```

The most relevant output often is sel_cv as it contains the results of the neutrality tests at gene level.
The globaldnds output has a table with global MLEs for the dN/dS ratios across all genes and their confidence intervals.
The annotmuts output contains a table with annotated coding mutations.
genemuts has a table with observed and expected number of mutations per gene.

## Table of significant genes

dout\$sel_cv contains the results for all the analyzed genes.
**We can look at the significant genes in the table by filtering for genes with a qglobal_cv \< 0.1.**

```{r}
dout$sel_cv[which(dout$sel_cv$qglobal_cv<0.1),]
```


```{r}
# Coeficiente de selección para mutaciones missense en SMAD4
wmis_SMAD4 = dout$sel_cv$wmis_cv[dout$sel_cv$gene_name == "SMAD4"]
n_mis_SMAD4 = dout$sel_cv$n_mis[dout$sel_cv$gene_name == "SMAD4"]

# Proporción de mutaciones seleccionadas
selected_prop = (wmis_SMAD4 - 1) / wmis_SMAD4 
selected_muts = n_mis_SMAD4 * selected_prop
cat("Mutaciones missense bajo selección positiva en SMAD4:", round(selected_muts), "\n")

# Confianza del coeficiente
geneci(dout, gene_list = "SMAD4")
```

```{r}
# Mutaciones observadas vs esperadas en SMAD4
dout$genemuts[which(dout$genemuts$gene_name == "SMAD4"), ]
```


```{r}
# Extraer todas las mutaciones codificantes observadas en el gen TP53
dout$annotmuts[which(dout$annotmuts$gene=="TP53"),]
```

Look at the aachange column to see the amino acid changes generated by the mutations.

```{r}
mutaciones_tp53 = dout$annotmuts[which(dout$annotmuts$gene=="TP53"),]
table(mutaciones_tp53$aachange)
```

```{r}
# Ver d eforma mas organizada lo de arriba! 
# Crear tabla de frecuencias
freqs <- table(mutaciones_tp53$aachange)

# Convertir a data frame y ordenar de mayor a menor frecuencia
df_freqs <- as.data.frame(freqs)
colnames(df_freqs) <- c("Cambio_Aminoacido", "Frecuencia")
df_freqs <- df_freqs[order(-df_freqs$Frecuencia), ]

# Mostrar de forma clara
print(df_freqs)
```

## Global signals of selection

dndscv also estimates global dN/dS ratios in aggregate for all the genes.
You can access this data in the dndscv output:

```{r}
print(dout$globaldnds)
```


## We can use the globaldnds information to estimate the number of missense driver mutations per sample.

```{r}
w = dout$globaldnds[1,2]
n_mis = length(which(dout$annotmuts$impact=="Missense"))
num_samples = length(unique(muts_dndscv$sampleID))

# Proporción seleccionada positivamente
prop_sel = (w - 1) / w
sel_mutations = prop_sel * n_mis

# Cálculo de mutaciones drivers en promedio por muestra
mut_drivers_por_muestra <- sel_mutations / num_samples

print(paste("Proporción de mutaciones missense bajo selección positiva:", round(prop_sel, 4)))
print(paste("Mutaciones drivers por muestra:", round(mut_drivers_por_muestra, 2)))
```

## Site/Codon level signals of selection

### Analysis of hotspots

We will now look for signals of positive selection at specific DNA or protein sites.
Firstly, have a look at the annotmuts output and try to determine by eye if there are hotspots. A couple lines of code which may help with the task:

```{r}
dout$annotmuts$gene_and_aachange = paste(dout$annotmuts$gene,
                                         dout$annotmuts$aachange, 
                                         dout$annotmuts$ntchange, 
                                         dout$annotmuts$pos,
                                         dout$annotmuts$impact, 
                                         sep=":")
sort(table(dout$annotmuts$gene_and_aachange),decreasing=T)[1:10]
```


## Using sitednds: selection at speciﬁc sites
### Running sitednds

```{r}
# Load the the Cancer Gene Census (v81) genes
data("cancergenes_cgc81", package="dndscv")
dout_cancergenes = dndscv(muts_dndscv, outmats=T, gene_list=known_cancergenes)
```



```{r}
sout = sitednds(dout_cancergenes)
```

```{r}
names(sout)
```

Now we are going to analyze recurrently mutated sites (potential hotspots) in cancer genes that are under significant positive selection.

```{r}
sout$recursites[which(sout$recursites$qval<0.1),]
```
```{r}
# Hacemos que la tabla a terior se vea mas ordenada
genes_significativos = dout$sel_cv[which(dout$sel_cv$qglobal_cv<0.1),]
hotspots_significativos = sout$recursites[which(sout$recursites$qval<0.1),]

# Ordenar por mayor significancia
hotspots_significativos = hotspots_significativos[order(hotspots_significativos$qval), ]

library(ggplot2)

# Gráfico de los genes más significativos
if(nrow(genes_significativos) > 0) {
  top_genes <- head(genes_significativos[order(genes_significativos$qglobal_cv), ], 20)
  
  ggplot(top_genes, aes(x = reorder(gene_name, -log10(qglobal_cv)), 
                       y = -log10(qglobal_cv))) +
    geom_bar(stat = "identity", fill = "steelblue") +
    coord_flip() +
    labs(title = "Driver genes in lung squamous cell carcinoma (LUSC)",
         x = "Gen",
         y = "-log10(q-value)") +
    theme_minimal()
}
```

## Predicting drivers in a given donor using the Cancer Genome Interpreter

We will use the Cancer Genome Interpreter to predict drivers in one of our donors. To make it more interesting, each one can select one donor randomly:


```{r}
# Seleccionar un sample aleatorio del conjunto muts_dndscv
random_donor = sample(unique(muts_dndscv$sampleID), 1)

muts_in_random_donor = muts_dndscv[which(muts_dndscv$sampleID == random_donor),
                                   c("chr", "pos", "ref", "mut")]
cat(random_donor, "donor has", nrow(muts_in_random_donor), "mutations\n")

```


**Preparing the input file**

Then, this code saves the mutations of the previously selected random donor (sample) into a `.tsv` file (tab-separated values).

```{r}
#  Guardar mutaciones en formato TSV
write.table(
  muts_in_random_donor,                         # DataFrame con las mutaciones del donante
  file = paste(random_donor, "_mutations.tsv", sep = ""),# Nombre del archivo basado en el ID del donante (ej. "Sample123.tsv")
  col.names = FALSE,                            # No escribe los nombres de columnas
  row.names = FALSE,                            # No escribe los índices de filas
  quote = FALSE                                 # No pone comillas alrededor de los valores
)

# Mensaje con instrucciones
cat("\nArchivo generado:", paste0(random_donor, "_mutations.tsv"), "\n")
cat("Por favor sube este archivo a: https://www.cancergenomeinterpreter.org/analysis\n")
cat("Asegúrate de seleccionar hg19 como genoma de referencia\n")

```

